### ISSUE TITLE
Echo of the of SSG shot became much quieter.

### ISSUE THREAD

**Issue by AntonieS** [dhewm/dhewm3#575]
OpenAL Soft 1.23.1, Win7 x64 SP1. Dhewm 3 1.5.3.1305.
Subj compare to 1.5.2.1503. Now it sounds unnatural.

**Comment by DanielGibson** [comment]
How does it sound compared to 1.4.1?
In 1.5.0 I reduced the volume of all sounds, which (due to some sounds being configured much louder than the can be played, and the max volume being clamped before setting it in OpenAL) changed the relative volume of some sounds, see https://github.com/dhewm/dhewm3/issues/326#issuecomment-1366833004 for an explanation.

In 1.5.3 that is fixed (I clamp the volume before reducing it), so it *should* sound like in 1.4.1 and older (incl. orig Doom3 with OpenAL backend) again, except for the bug that was fixed by reducing the sounds in the first place (https://github.com/dhewm/dhewm3/issues/179), which is still fixed.

**Comment by AntonieS** [comment]
1.4.1 is very bad.
1.5.2 sounds much better. But its about echo. 
Overall 1.5.3 lesser SSG echo quality but better in other: some ambient sounds, for example announcements that the passage is closed in Admin map, appears only in 1.5.3.

**Comment by tyuah8** [comment]
In my opinion, the code written to reduce the volume of the sounds here:

[4567f26](https://github.com/dhewm/dhewm3/commit/4567f265392b5ab95a335f8896ff003a743990e4)

is both unnecessary and incorrect, as all it does is move the problem to a different volume, since it multiples the volume by 0.333f.

I recommend **completely restoring the original game code (not just removing the 0.333f)**, and instead, use the following in alsoft.ini:

## output-limiter:
#  Applies a gain limiter on the final mixed output. This reduces the volume
#  when the output samples would otherwise clamp, avoiding excessive clipping
#  noise.
**output-limiter** = **true**

By default, **output-limiter** is set to **false**, which causes the situation where too many loud sounds playing causes some of them to go quiet.  Setting it to **true** completely eliminates this issue.

**Comment by DanielGibson** [comment]
I don't have influence on users alsoft.ini/alsoft.conf

**Comment by tyuah8** [comment]
Very well, but perhaps you can make your changes optional so users have the option of doing what I just mentioned (using **output-limiter** = **true** instead of the hard-coded volume reduction, which seems to be causing other issues).

It is what I opted for with my version of the DOOM 3 engine.

**Comment by DanielGibson** [comment]
Yeah, that would be an option

**Comment by DanielGibson** [comment]
```
## output-limiter:
#  Applies a gain limiter on the final mixed output. This reduces the volume
#  when the output samples would otherwise clamp, avoiding excessive clipping
#  noise.
#output-limiter = true
```

Are you sure that setting it to **true** would help?

To me it looks like true is the default, and if anything, then setting it to false might help, at least with the problem I fixed with the initial `*= 0.333` fix (lots of loud sounds lowering overall volume and thus making other sounds inaudible, what's described in https://github.com/rorgoroth/dhewm3/commit/3c01757d27a22a3858737f6b51e85895d2fac887 and #179)

**Comment by tyuah8** [comment]
Yes it would help.  I tested this, and it is counter-intuitive, but **false** is the default contrary to what the documentation says.

You can test this very easily if you set it to **false**, and run Dhewm3 with s_volume_dB "10"; you will find that when you fire the shotgun at a wall, the shotgun sound is muffled by the bullets hitting the wall, but the moment you set **output-limiter** = **true**, the problem goes away.

Your fix actually does not fix the problem, as you apply both the clamp and scale **before** the s_volume_dB, so all the user has to do to "undo your fix" is increase the volume, which he would have to do, since everything is now quieter; the user would increase by 10 dB, since 10 dB is roughly x3, which counters your 0.333

**Comment by DanielGibson** [comment]
> the user would increase by 10 dB, since 10 dB is roughly x3, which counters your 0.333.

that's their problem then, not sure what I should do about it. My goal is for it to sound ok at the default volume, as that's what most people use - and keep in mind that `s_volume_dB = 0` is the value you get when turning the slider in doom3's system menu all the way to the right, higher values like 10 are only possible through the console.

Fact is, doom3's audio is inherently buggy, due to some sounds having a way too loud gain configured that only sounds ok-ish due to being clamped to 1.0 (this is the issue I fix by clamping to 1 before scaling down), so whatever I do, it'll be hard to properly fix the issue in a way that always sounds "good" (if Doom3 even ever sounded good in all those edge-cases that cause trouble now) an doesn't depend on OpenAL hopefully handling too many loud sounds that are played at the same time in a way that doesn't sound too bad, or at least behaves similarly to Creative's old OpenAL implementation, or whatever the benchmark is.

(The "proper" way would be editing all sound shaders, basically remixing the game's sound, which I don't really want to do, and I'm not qualified for anyway)

**Comment by tyuah8** [comment]
Well, it is your repo, so you decide what makes sense and doesn't, but what I am doing for my DOOM 3 is the following:

- use **output-limiter** = **true**, which fixes the sound muffling issue without any source code change
- manually edit every single sound volume that I personally feel is either too quiet or too loud, and I am doing this by modifying the .map file and/or the .sndshd file

**Comment by DanielGibson** [comment]
I'll look into it..

by the way, what project is "your Doom3"? :)

**Comment by tyuah8** [comment]
I have the source code locally, but I original forked from [TTimo](https://github.com/TTimo/doom3.gpl) 12 years ago to work on something widescreen related.  I recently went back to working on DOOM 3 with new changes to both the engine and the pk4 files.

One of the changes you made that I really liked was the fixes to OpenAL and [296](https://github.com/dhewm/dhewm3/pull/296) which I copied over to my source code.  I gave you credit by referencing the Pull Request and or commits involved.

So... excellent work to you!

**Comment by DanielGibson** [comment]
Your suggestions (together with a menu that allows configuring this and even OpenALs output-limiter) is implemented in https://github.com/DanielGibson/dhewm3/tree/imgui-rebased

Can't say that I heard much of a difference with the double shotgun though

**Comment by j4reporting** [comment]
> ```
> #output-limiter = true
> ```
> 
> To me it looks like true is the default, 

It sure looks like this, but  the debug log suggest this is not the case.

```
[ALSOFT] (II) Key output-limiter not found
[ALSOFT] (II) Output limiter disabled
```

This may be a bug, because the included openal-config  program also assumes that  the output-limiter (Gain Limiter)  is  enabled by default. 
The GUI (at least  the Windows version)  does not write the ```output-limiter=true```  line in the ```%APPDATA%\alsoft.ini ``` file.  Only if disabled then the line output-limiter will be set to false in the alsoft.ini file.


You can also override the output-limiter setting in a local  alsoft.ini file placed in the dhewm3 directory. 
It will only override the settings included in the game's alsoft.ini  file. 

i.e.  ( had some settings configured in %APPDATA% )
```
[General]
stereo-mode=headphones
default-hrtf=Built-In HRTF
excludefx="eaxreverb,reverb"
```

here's the openal debug log  (trimmed down)

```
[ALSOFT] (II) Initializing library v1.23.1-d3875f33 master
[ALSOFT] (II) Supported backends: wasapi, dsound, winmm, null, wave
[ALSOFT] (II) Loading config C:\Users\<username>\AppData\Roaming\alsoft.ini...
[ALSOFT] (II)  found 'stereo-mode' = 'headphones'
[ALSOFT] (II)  found 'default-hrtf' = 'Built-In HRTF'
[ALSOFT] (II)  found 'excludefx' = 'eaxreverb,reverb'
[ALSOFT] (II) Got binary: c:\Games\dhewm3-pre, dhewm3.exe
[ALSOFT] (II) Loading config c:\Games\dhewm3-pre\alsoft.ini...
[ALSOFT] (II)  found 'output-limiter' = 'true'
....
[ALSOFT] (II) Found excludefx = "eaxreverb,reverb"
[ALSOFT] (II) Key default-reverb not found
[ALSOFT] (II) Key eax/enable not found
[ALSOFT] (II) EAX disabled because EAXReverb is disabled.
...
[ALSOFT] (II) Found stereo-mode = "headphones"
[ALSOFT] (II) Found default-hrtf = "Built-In HRTF"
[ALSOFT] (II) Loading !1_Built-In HRTF...
....
[ALSOFT] (II) Found output-limiter = "true"
[ALSOFT] (II) Output limiter enabled, 0.0000dB limit

```

> Can't say that I heard much of a difference with the double shotgun though

The best way to hear it is to take 1-2 steps back from the wall.

**Comment by DanielGibson** [comment]
In the branch I linked above you can enable and disable the output limiter in dhewm3, with a cvar or by clicking in the new settings menu.

It indeed seems to be disabled by default.

**Comment by j4reporting** [comment]
yeah, I just tried to compile the branch with VStudio  2022  but I get these errors:

```
1>Dhewm3SettingsMenu.cpp
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstdio(82,25): error C2039: 'use_idStr_snPrintf': is not a member of '`global namespace''
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstdio(82,13): error C2873: 'use_idStr_snPrintf': symbol cannot be used in a using-declaration
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstdio(83,25): error C2039: 'use_idStr_vsnPrintf': is not a member of '`global namespace''
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstdio(83,13): error C2873: 'use_idStr_vsnPrintf': symbol cannot be used in a using-declaration
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstring(32,25): error C2885: 'idStr::Cmp': not a valid using-declaration at non-class scope
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstring(39,25): error C2039: 'use_idStr_Cmpn': is not a member of '`global namespace''
1>C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.40.33807\include\cstring(39,13): error C2873: 'use_idStr_Cmpn': symbol cannot be used in a using-declaration

```

**Comment by DanielGibson** [comment]
Try again, I just pushed a fix for that :)

**Comment by DanielGibson** [comment]
why does it use libz.a, shouldn't it use zlib1.lib?

**Comment by j4reporting** [comment]
good catch.  /SAFESEH  was always  enabled, but
it seems the cmake FindZLIB.cmake module has changed
not sure what enabled  the search for static libs. 

 CMAKE 3.26.4
-- Found ZLIB: C:/Users/XXXX/source/repos/dhewm3-libs/i686-w64-mingw32/lib/**libz.dll.a** (found version "1.2.7")

CMAKE  3.29.3
-- Found ZLIB: C:/Users/XXXX/source/repos/dhewm3-libs/i686-w64-mingw32/lib/**libz.a** (found version "1.2.7")

**Comment by DanielGibson** [comment]
nevermind, I just pushed a commit that removes zlib and uses integrated [miniz](https://github.com/richgel999/miniz/) instead

**Comment by j4reporting** [comment]
regarding the openal config GUI utility.  The checkbox for 'Gain Limiter'  has 3 options!   Default, enabled  and disabled. 
The problem here is that on Windows you can't see any difference between  default and enabled  state! 

on Linux  the 'default' option  is unchecked  and has a dark grey color.

![alsoft_gui_default](https://github.com/dhewm/dhewm3/assets/45177907/8963866b-2fc9-4cf2-9674-6668490573bc)

**Comment by DanielGibson** [comment]
You should report that to [OpenAL Soft](https://github.com/kcat/openal-soft/), I'm not involved with that project :)

In dhewm3 the three states should be easy enough to tell apart
![image](https://github.com/dhewm/dhewm3/assets/1216493/df176d11-fa9c-406a-ace8-3fe22866e298)

**Comment by DanielGibson** [comment]
Here's a current build for Windows: 
[dhewm3-1.5.4pre-imgui3_win32.zip](https://github.com/user-attachments/files/15546318/dhewm3-1.5.4pre-imgui3_win32.zip)

**Comment by j4reporting** [comment]
yeah no issues  with the output-limiter setting in imgui.

One point regarding the imgui, though.
On first start of the imgui  the window should be resized to something more sane  than  
Size=32,72

the default  size is ignored

here is my  imgui.inf  after opening and closing the imgui for first time:

```
[Window][Debug##Default]
Pos=60,60
Size=400,400

[Window][dhewm3 Settings]
Pos=60,60
Size=32,72
```

**Comment by Kappa971** [comment]
Not sure what the problem was with this fix but this patch https://github.com/bibendovsky/eaxefx/pull/28#issuecomment-1367252180 for the original game works well and allows you to decrease the volume from the game menu without changing the "balances" of the sounds, and does not rely on OpenAL or anything else but the game "solves" a problem caused by bad sound adjustment by the developers.
Also the gain limiter applied here, compared to the patch indicated above, could cause sudden drops in volume in cases where the game volume would be too high.
